design/open/lakefs-on-kv-testing-plan.md

```diff
@@ -0,0 +1,128 @@
+Updates are done for commitID, as part of `post` hooks
+* Currently unit tests cover hooks run and verify the results in DB using the service reading functions. There are also error testings (hook failures)
+* 67.6% coverage
+* What is missing:
+  * DB Failure during hook action (TBD - what is the expected behavior? Is that interesting to test?)
+  * Performance under load - not sure a load on actions/hooks is relevant
+  * Concurrent actions/hooks execution, which derives concurrent DB accesses. This gets more interesting if load is involved (in case load is relevant)
+
+### ```pkg/auth```
+* DB is used to store installation metadata, users and policies
+* Only 6.8% coverage
+* Some benchmarks exists that covers effectivePolicies (multi-table JOIN)
+* What is missing:
+  * Need to increase code coverage
+  * Comparison benchmark to toggle between Tables and KV and verify there is no degradation
+  * Migration tests:
+    * verify that data is migrated correctly
+    * Authentication transactions in during migration (data that was written to tables is read from KV)
+
+### ```pkg/diagnostics```
+* Only read access
+* No test coverage. Need to verify all is working the same after migration
+
+### ```pkg/graveler/ref```
+* Branch locking (read)
+* Repos, branches, commits and tags - read/write
+* 95.5% testing coverage
+* No performance/benchmarking
+* Missing:
+  * Migration tests
+  * Comparison benchmark
+
+### ```pkg/graveler/retention```
+* No direct DB access. Accesses are done using ```pkg/graveler/ref```, which should make the DB transition seamless
+* 42.9% coverage in unit testing
+* Missing performance tests that might help detect degradation
+
+### ```pkg/graveler/staging```
+* 89.1% coverage
+* Covers all staged KV functionality
+* No performance/benchmarking
+* Missing:
+  * Migration tests
+  * Comparison benchmark
+
+### Res-Dump/Restore
+Currently this functionality is not covered directly, but since it relies on '''pkg/graveler/ref''' for DB access, the DB migration should be seamless
+It can be leveraged, however, to extend the cover of '''pkg/graveler/ref''' and for performance, as it is quite exhaustive (traverses all branches and commits and tags, per repo)
+
+## Execution Plan
+
+### KVM1
+* Data level migration tests infrastructure
+  * Migrate data from Table to KV, dumps both and compares
+  * Implement dumpers for `gateway_multiprts`
+  * Implement tests for '''pkg/multiparts'''
+* Infrastructure for running migration during a system test execution
+  * System test to run migration during multiparts upload
+  * Currently there is a single simple multiparts system test (single file, 7 parts) - this is also an opportunity to expand that
+* KV Store unit tests 
+* `multiparts.Tracker` benchmark to run on both Table DB and KV Store (use feature flag to toggle) and verify there is no degradation
```

Comment:
Should we run it once locally or as part of nessie/something else?

Reply:
This is intended to be a unit-test, so outside of `Esti` context. It could be used locally for developement/debugging purposes, but obviously, I would like it to run as part of our '''Go Tests''' and fail if an unacceptable degradation is detected

Reply:
> if an unacceptable degradation is detected

How? It depends on your env where you run the test, no? I think benchmarking should be executed manually by us before releasing and checking it as we continue to develop the feature.

Reply:
I think that the maintenance burden of performance testing as a gatekeeper is huge. It will sporadically fail, require a lot of tuning and generally really hard to get right. And since we agree that it's temporary testing (until we complete the feature), I think we'll get much more value from a local execution that we control and can run before each feature/milestone release.